common/src/main/java/io/pravega/common/io/serialization/RevisionDataOutputStream.java

```diff
@@ -0,0 +1,279 @@
+        } else if (value > 0x3FFF_FFFF) {
+            // All 8 bytes
+            writeInt((int) (value >>> 32 | 0xC000_0000));
+            writeInt((int) value);
+        } else if (value > 0x3FFF) {
+            // Only 4 bytes.
+            writeInt((int) (value | 0x8000_0000));
+        } else if (value > 0x3F) {
+            // Only 2 bytes.
+            this.out.write((int) (value >>> 8 | 0x40 & 0xFF));
+            this.out.write((int) (value & 0xFF));
+        } else {
+            // 1 byte.
+            this.out.write((byte) value);
+        }
+    }
+
+    @Override
+    public int getCompactIntLength(int value) {
+        if (value < COMPACT_INT_MIN || value >= COMPACT_INT_MAX) {
+            throw new IllegalArgumentException("writeCompactInt can only serialize non-negative longs up to 2^30.");
+        } else if (value > 0x3F_FFFF) {
+            return 4;
+        } else if (value > 0x3FFF) {
+            return 3;
+        } else if (value > 0x3F) {
+            return 2;
+        } else {
+            return 1;
+        }
+    }
+
+    @Override
+    public void writeCompactInt(int value) throws IOException {
+        if (value < COMPACT_INT_MIN || value >= COMPACT_INT_MAX) {
+            throw new IllegalArgumentException("writeCompactInt can only serialize non-negative longs up to 2^30.");
+        } else if (value > 0x3F_FFFF) {
+            // All 4 bytes
+            this.out.write(value >>> 24 | 0xC0 & 0xFF);
+            this.out.write(value >>> 16 & 0xFF);
+            this.out.write(value >>> 8 & 0xFF);
+        } else if (value > 0x3FFF) {
+            // 3 bytes.
+            this.out.write(value >>> 16 | 0x80 & 0xFF);
+            this.out.write(value >>> 8 & 0xFF);
+        } else if (value > 0x3F) {
+            // 2 Bytes.
+            this.out.write(value >>> 8 | 0x40 & 0xFF);
+        }
+
+        // Last byte is always the same.
+        this.out.write(value & 0xFF);
+    }
+
+    @Override
+    public void writeUUID(UUID uuid) throws IOException {
+        writeLong(uuid.getMostSignificantBits());
+        writeLong(uuid.getLeastSignificantBits());
+    }
+
+    @Override
+    public <T> void writeCollection(Collection<T> collection, ElementSerializer<T> elementSerializer) throws IOException {
+        if (collection == null) {
+            writeCompactInt(0);
+            return;
+        }
+
+        writeCompactInt(collection.size());
+        for (T e : collection) {
+            elementSerializer.accept(this, e);
+        }
+    }
+
+    @Override
+    public <K, V> void writeMap(Map<K, V> map, ElementSerializer<K> keySerializer, ElementSerializer<V> valueSerializer) throws IOException {
+        if (map == null) {
+            writeCompactInt(0);
+            return;
+        }
+
+        writeCompactInt(map.size());
+        for (Map.Entry<K, V> e : map.entrySet()) {
+            keySerializer.accept(this, e.getKey());
```

Comment:
shouldnt we write size of each key and value into the serialization of map?
because keySerializer and keyDeserializer could be working with different revisions to serialize and deserialize.

Reply:
That is the responsibility of the key/value serializer implementation. You can pass in instances of `VersionedSerializer` to this, which will do that.

Check out the unit tests for this class, they cover this scenario.

Reply:
Yes, that is true. Element serializer here can be:
1. A function in `DataOutput` (for primitive types)
2. A `VersionedSerializer.serialize` for another complex type with a serializer
3. A user-defined custom method.

1 and 2 are OK as long as the deserialization is done by respecting the format. That is, if you use `writeUTF`, you need to call `readUTF`, or if you use `Foo.SERIALIZER::serialize` you need to use `Foo.SERIALIZER::deserializer`. 

If the user defines their own thing here (3 above), they are on the hook for providing a proper deserialization method. However I strongly discourage this approach as you can achieve the same thing by writing a new `VersionedSerializer` which takes care of all of these issues. There are some special cases, though, when using a dedicated serializer may not be suitable; in those cases you may want to write them yourself.

Check out what I did for the Segment Store (no PR yet, but the branch is there). For `StreamSegmentAppendOperation`, I need to encode a collection of AttributeUpdates. I could have created a serializer for these AttributeUpdates, but I chose not to, because that would add too much overhead. This was a deliberate choice: https://github.com/pravega/pravega/blob/issue-2310-store-serialization/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/operations/StreamSegmentAppendOperation.java#L153

Reply:
yes, the concern is with future maintainability of the code when OSS developers start contributing. I am wondering if we can handle some of the things in the framework which prevents developers from making such mistakes. 
The problem really comes with potential version mismatch when the map data types are changed and they were not encoded with a versioned serializer. So there is no information that really tells us while reading a serialized blob how it is encoded. 
Say for example you have a collection<int> which is later changed to collection<String>.
we update our elementSerializers but the encoded serialized data from previous generation (when collection was of `integer`) has no information in it to say it is incompatible with String. 
Unless we force wrapping of int and  strings inside some versioned serializer.